src/transaction/transaction.ts

```diff
@@ -1,45 +1,128 @@
 import * as InteropLib from 'cardano-hw-interop-lib'
-import { isEqual, uniqWith } from 'lodash'
-import { encodeTx, RawTransaction } from 'cardano-hw-interop-lib'
+import { isEqual, uniqWith, cloneDeep } from 'lodash'
 import {
   TxWitnessByron,
   TxWitnessShelley,
   TxCborHex,
   TxWitnessKeys,
 } from './types'
 import { encodeCbor } from '../util'
+import { CardanoEra } from '../types'
+import { SigningParameters } from '../crypto-providers/types'
+
+type WitnessSet = Map<number, unknown[]>
+
+// If witnessSet is empty, the CBOR lib parses it as {}, otherwise it seems to always return a Map.
+// Let's convert it to Map in any case, just to make sure.
+const _parseWitnessSet = (witnessSet: unknown): WitnessSet => {
+  const clonedWitnessSet = cloneDeep(witnessSet)
```

Comment:
I added the deep cloning because I don't like the idea of `TxSigned` constructor modifying the tx witness set that it received, whet it was supposed to take the parameters and create a new assembled tx.

> or generally always construct a new Map as in the last line

Without the explicit deep cloning, this would be only a shallow copy.